packages/graphql-hooks/index.d.ts

```diff
@@ -185,6 +185,7 @@ export interface UseClientRequestOptions<
   fetchOptionsOverrides?: object
   updateData?(previousData: ResponseData, data: ResponseData): any
   client?: GraphQLClient
+  responseReducer?(response: object): object
```

Comment:
We can be more specific here:

```diff
- responseReducer?(response: object): object
+ responseReducer?(response: Response): object
```

The `UseClientRequestResult` type also needs to be updated with the new key.

Reply:
Hi @gdorsi thanks again. Does `Response` refer to the Fetch API? if so we shouldn't change it since graphql-hooks allow to provide other fetch implementations which then may have another signature, right?

Reply:
uh, thats true!

You have done right XD

Reply:
Hi @gdorsi 👋 
Yes just had a look at it and from MPOV we should already be fine with the current state.
The `request` fns uses generics as well but defaults to `any` if not specified 👍 
![Bildschirmfoto 2021-12-16 um 10 38 00](https://user-images.githubusercontent.com/3241476/146346460-c1fa9dc2-930a-4875-9436-3eb010b18ece.png)

so i made a quick check again on the TS example in the repo and passed in a custom generic for the responseData (resolve case of the promise):
![Bildschirmfoto 2021-12-16 um 10 45 36](https://user-images.githubusercontent.com/3241476/146347785-8445299d-30f6-487b-a80b-2c1f7b6883de.png)

or did you mean smth different?
thanks again 🙏

Reply:
I'm talking about this:

<img width="726" alt="Screenshot 2021-12-16 at 16 50 10" src="https://user-images.githubusercontent.com/6529568/146403925-8b445281-8ac6-422d-bbce-6d5fb2ef9e0c.png">

Or I'm doing something wrong?

Reply:
Hi @gdorsi 
Thanks once again 🙏  now i understand what you've meant 😅 

Guess i'd have two options to solve this:
- return the `responseReducer` data within the `data` key of the response under an unique key (`__responseReducer` for example) to avoid collisions with fetched data. the generic in this case could look smth like this then: `client.request<{data: {some: string, cool: string, stuff: any}, __responseReducer: { contentType: string}}, any, any>`
- add a 4th generic to all methods which return `Result<ResponseData, TGraphqlError>` so that we can type everything there. example could look like `client.request<{data: {some: string, cool: string, stuff: any}}, any, any, { contentType: string, contentLength: number}>`

or do you have any other in mind? which one would you prefer?

<hr />

**random question**
when i've played around with the 2nd version i've realized that as soon as i change the `ResponseData` type from `any` to a specific value, the `data` key is set optional so i would have to use `data?.myValue` to access it. 

![Bildschirmfoto 2021-12-17 um 22 11 47](https://user-images.githubusercontent.com/3241476/146608977-cb0df99a-183e-439a-9662-a25b5d875759.png)

Comes from L175 in the `index.d.ts` file. 
![Bildschirmfoto 2021-12-17 um 22 11 31](https://user-images.githubusercontent.com/3241476/146608964-7eb9a93c-134b-4261-98dc-e1b767e66afa.png)

How do you deal with that when you specify the data returned from the server? check if an error occurred and if not, cast the response before using it?
![Bildschirmfoto 2021-12-17 um 22 28 33](https://user-images.githubusercontent.com/3241476/146610668-9e5efd45-746c-4dfe-af08-dceba3e8b46d.png)

Reply:
> or do you have any other in mind? which one would you prefer?

I'm wondering, what if we change the behavoir of `responseReducer` and let the user enrich the `data` field with some information from the response?

Something like this:
```js
const { data } = await client.request({ query }, {
 responseReducer: (response, data) => ({
   ...data,
   cache: response.headers.get('x-cache-tags')
 })
})
```

What do you think?

Reply:
thank you @gdorsi 👍  i've implemented that one now